### ISSUE TITLE
Add sample-size-weighted fusion regularization

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/multidms#198]
# Add sample-size-weighted fusion regularization

## Summary

Add a `scale_fusion_by_n` parameter that weights each condition's fusion (shift lasso) penalty by `n_ref / n_d`, where `n_ref` is the reference condition's variant count and `n_d` is the non-reference condition's count. This corrects an asymmetry where data-poor conditions receive disproportionately strong shrinkage relative to their information content, without introducing any additional hyperparameter sweep dimensions.

## Background

- **Current behavior:** The fusion penalty `λ × Σ_d |β_d − β_ref|₁` applies the same global strength λ to all non-reference conditions. Meanwhile, the loss function (`functional_score_loss`) uses `.sum()` over variants, so conditions with more data naturally contribute more to the total loss. This creates an imbalance: data-rich conditions dominate the loss gradient but share the same regularization pressure as data-poor conditions.
- **Why this is limiting:** In the SARS-CoV-2 spike analysis, Delta has ~150K variants while BA.2 has ~308K (~2× more). Both have shifts relative to BA.1 (reference, ~232K variants), but Delta shifts have higher intrinsic variance (stdev 0.447 vs 0.356) and are harder to estimate reliably. The uniform fusion penalty over-shrinks Delta's genuine shifts — the same λ=4.0 that appropriately regularizes BA.2 may be too aggressive for Delta given its lower sample size.
- **What the fusion penalty does:** It's the L1 (lasso) penalty on `β_d − β_ref`, implemented via a proximal gradient operator (`jaxopt.prox.prox_lasso`). This penalty encourages sparsity in shifts — most mutations should have the same effect across conditions, with only a few showing condition-specific differences. The proximal operator applies soft-thresholding with threshold `λ / scale`.

## Proposed Approach

Weight each condition's fusion penalty by `n_ref / n_d`. When a condition has fewer variants than the reference, its effective fusion strength decreases proportionally — reflecting the fact that its shift estimates are noisier and should not be shrunk as aggressively. When `n_d = n_ref`, the weight is 1.0 (no change). When `n_d > n_ref`, the weight is > 1.0 (slightly stronger shrinkage for data-rich conditions, which is also sensible since their estimates are more reliable and can tolerate more regularization).

**Key design decisions:**
- **Boolean flag, not a continuous parameter** — `scale_fusion_by_n=True/False` rather than a user-specified weight dict. The weights are derived entirely from the data, so there's nothing to tune. This avoids adding sweep dimensions.
- **Reference condition as the anchor** — weights are `n_ref / n_d`, not `max(n) / n_d` or `mean(n) / n_d`. This matches the model's asymmetric structure where shifts are defined relative to the reference condition.
- **Backward-compatible** — defaults to `False`, preserving current behavior for all existing code.

### Programmatic API

```python
import multidms

# Current behavior (unchanged):
model = multidms.Model(data, fusionreg=4.0)
model.fit()

# New: sample-size-weighted fusion
model = multidms.Model(data, fusionreg=4.0, scale_fusion_by_n=True)
model.fit()

# With ModelCollection parameter sweeps:
params = {
    "dataset": [data],
    "fusionreg": [0.0, 2.0, 4.0, 8.0],
    "scale_fusion_by_n": [True],
}
models = multidms.fit_models(params)
```

### Verbose output

When `verbose=True`, the effective per-condition fusion weights are printed once at the start of fitting:

```
Fusion weights (scale_fusion_by_n=True):
  Delta: 1.55 (n_ref=231934, n_d=149616)
  Omicron_BA2: 0.75 (n_ref=231934, n_d=307886)
```

### 1. `fit()` in `jaxmodels.py`

**File:** `multidms/jaxmodels.py` (existing, line 354)

Add `scale_fusion_by_n` parameter and compute per-condition fusion weights from variant counts:

```python
def fit(
    data_sets: dict[str, Data],
    reference_condition: str,
    l2reg: Float = 0.0,
    fusionreg: Float = 0.0,
    beta0_ridge: Float = 0.0,
    scale_fusion_by_n: bool = False,  # NEW
    ...
) -> tuple[Model, pd.DataFrame]:
```

Compute weights early in `fit()`, before the optimization loop:

```python
# Compute per-condition fusion weights
n_ref = data_sets[reference_condition].X.shape[0]
if scale_fusion_by_n:
    fusion_weights = {
        d: n_ref / data_sets[d].X.shape[0]
        for d in data_sets
        if d != reference_condition
    }
else:
    fusion_weights = {
        d: 1.0 for d in data_sets if d != reference_condition
    }
```

### 2. `objective_total` in `jaxmodels.py`

**File:** `multidms/jaxmodels.py` (existing, line 459)

The `fusion_weights` dict is captured by closure (like `loss_fn` and `loss_kwargs`). Modify the fusion penalty accumulation:

```python
@jax.jit
def objective_total(
    model, data_sets, l2reg=0.0, fusionreg=0.0, scale=1.0, beta0_ridge=0.0
):
    loss = sum(loss_fn(model, data_sets, **loss_kwargs).values())
    l2_penalty = 0.0
    fusion_penalty = 0.0
    for d in data_sets:
        β = model.φ[d].β
        l2_penalty += (β**2).sum()
        if d != model.reference_condition:
            fusion_penalty += fusion_weights[d] * jnp.abs(
                model.φ[d].β - model.φ[model.reference_condition].β
            ).sum()
    return (
        loss
        + l2reg * l2_penalty
        + fusionreg * fusion_penalty
        + _beta_ridge_penalty(model, beta0_ridge)
    ) / scale
```

Note: `fusion_weights` is captured from the enclosing `fit()` scope, just like `loss_fn` and `loss_kwargs` already are. The function signature does not change.

### 3. `prox_block` in `jaxmodels.py`

**File:** `multidms/jaxmodels.py` (existing, line 480)

Pass per-condition weights through the `hyperparameters` dict and apply them in the proximal lasso step:

```python
@jax.jit
def prox_block(β_block, hyperparameters, scaling=1.0):
    model = hyperparameters["model"]
    fusionreg = hyperparameters["fusionreg"]
    scale = hyperparameters["scale"]
    fusion_weights = hyperparameters["fusion_weights"]  # NEW
    beta_clip_range = hyperparameters.get("beta_clip_range", None)
    # lasso
    β_ref = β_block[model.reference_condition]
    for d in β_block:
        if d != model.reference_condition:
            β = β_block[d]
            Δ = β - β_ref
            Δ_lasso = jaxopt.prox.prox_lasso(
                Δ, fusion_weights[d] * fusionreg / scale, scaling
            )
            β_block[d] = β_ref + Δ_lasso
    # box clipping (if specified)
    if beta_clip_range is not None:
        clip_min, clip_max = beta_clip_range
        for d in β_block:
            β_block[d] = jnp.clip(β_block[d], clip_min, clip_max)
    return β_block
```

Update both `hyperparameters_prox` construction sites (lines 636 and 667) to include `fusion_weights`:

```python
hyperparameters_prox = dict(
    model=model,
    fusionreg=fusionreg,
    scale=scale,
    beta_clip_range=beta_clip_range,
    fusion_weights=fusion_weights,  # NEW
)
```

### 4. `Model.__init__()` in `model.py`

**File:** `multidms/model.py` (existing, line 58)

Add parameter and store it:

```python
def __init__(
    self,
    data: Data,
    loss_type: Literal["functional_score_loss", "count_loss"] = "functional_score_loss",
    ge_type: Literal["Identity", "Sigmoid"] = "Sigmoid",
    l2reg: float = 0.0,
    fusionreg: float = 0.0,
    beta0_ridge: float = 0.0,
    scale_fusion_by_n: bool = False,  # NEW
):
    ...
    self._scale_fusion_by_n = scale_fusion_by_n
```

### 5. `Model.fit()` in `model.py`

**File:** `multidms/model.py` (existing, line 226)

Thread through to `jaxmodels.fit()`:

```python
self._jax_model, self._convergence_trajectory_df = jaxmodels.fit(
    ...
    scale_fusion_by_n=self._scale_fusion_by_n,
    ...
)
```

### 6. `Model.__str__()` in `model.py`

**File:** `multidms/model.py` (existing, line 832)

Add `scale_fusion_by_n` to the display string.

### 7. `fit_one_model()` in `model_collection.py`

**File:** `multidms/model_collection.py` (existing, line 53)

Add `scale_fusion_by_n=False` parameter, pass to `Model()` constructor:

```python
def fit_one_model(
    dataset,
    ge_type="Sigmoid",
    l2reg=0.0,
    fusionreg=0.0,
    beta0_ridge=0.0,
    scale_fusion_by_n=False,  # NEW
    ...
):
    ...
    model = multidms.Model(
        dataset,
        ge_type=ge_type,
        l2reg=l2reg,
        fusionreg=fusionreg,
        beta0_ridge=beta0_ridge,
        scale_fusion_by_n=scale_fusion_by_n,  # NEW
        loss_type=loss_type,
    )
```

The parameter is automatically captured in `fit_attributes` via `locals().copy()` on line 118.

### Weight Formula

For non-reference condition d with `n_d` variants and reference condition with `n_ref` variants:

```
weight_d = n_ref / n_d
```

Example from the spike analysis (Omicron BA.1 as reference):
- **Delta**: weight = 231,934 / 149,616 ≈ **1.55** (less shrinkage)
- **Omicron BA.2**: weight = 231,934 / 307,886 ≈ **0.75** (more shrinkage)

### Interaction with the Proximal Operator

The `jaxopt.prox.prox_lasso(Δ, λ_eff, scaling)` applies soft-thresholding: `sign(Δ) * max(|Δ| - λ_eff * scaling, 0)`. By scaling λ_eff per condition, we adjust the soft-thresholding threshold independently for each condition. This is mathematically equivalent to solving the weighted group lasso:

```
min Σ_d loss_d + λ * Σ_d w_d * |β_d - β_ref|₁
```

where `w_d = n_ref / n_d`.

### JAX/JIT Compatibility

The `fusion_weights` dict maps condition names to Python floats. These are captured by closure in `objective_total` and passed as a plain dict in `hyperparameters` for `prox_block`. Since condition names are static (not traced) and weights are concrete floats, there are no JIT tracing issues.

## Testing Strategy

- **Unit test:** `test_scale_fusion_by_n_weights` — verify that with `scale_fusion_by_n=True`, a condition with fewer variants gets less shrinkage (more nonzero shifts) than with `scale_fusion_by_n=False`, at the same λ. Follow the `test_beta0_ridge_penalty` pattern: fit two models with/without the flag and compare shift sparsity.
- **Unit test:** `test_scale_fusion_by_n_equal_sizes` — when all conditions have equal variant counts, `scale_fusion_by_n=True` should produce identical results to `scale_fusion_by_n=False` (weights are all 1.0).
- **Unit test:** `test_scale_fusion_by_n_default_false` — verify the default is `False` and existing behavior is unchanged.
- **Integration test:** `test_fit_one_model_with_scale_fusion_by_n` — verify `fit_one_model()` correctly threads the parameter through and records it in `fit_attributes`.
- **Edge cases:**
  - Single non-reference condition (only one weight to compute)
  - `fusionreg=0.0` with `scale_fusion_by_n=True` (should be harmless — 0 × weight = 0)

## Documentation Updates

- [ ] Update `jaxmodels.fit()` docstring for `scale_fusion_by_n`
- [ ] Update `Model.__init__()` docstring
- [ ] Update `fit_one_model()` docstring

## Dependencies

None — no new dependencies required.

### Alternative 1: Per-condition fusion strengths (`fusionreg_per_condition` dict)

Allow users to specify separate λ values for each condition.

**Why not chosen:** Introduces a combinatorial hyperparameter sweep (e.g., 4 values × 2 conditions = 8 fits instead of 4). The sample-size weighting achieves the same goal automatically.

### Alternative 2: Normalize the loss by sample size (`.mean()` instead of `.sum()`)

Make the loss per-variant rather than total, which would equalize the gradient contribution across conditions.

**Why not chosen:** This is a more invasive change that affects the loss-regularization balance for all users and would require re-tuning hyperparameters. It's also a separate concern — the loss normalization and the regularization weighting address different aspects of the imbalance.

### Alternative 3: Adaptive lasso (two-stage fitting)

Fit once with uniform λ, then refit with weights `1/|Ŝ_d,j|` per mutation.

**Why not chosen:** More complex (two-stage) and addresses a different problem (adaptive per-mutation shrinkage vs. per-condition shrinkage). Could be a valuable future enhancement but is orthogonal to this proposal.

## Success Criteria

- [ ] `scale_fusion_by_n=False` (default) produces identical results to current code
- [ ] `scale_fusion_by_n=True` with equal-sized conditions produces identical results to `False`
- [ ] `scale_fusion_by_n=True` with unequal conditions produces less shrinkage for the smaller condition
- [ ] Parameter threads correctly through `Model`, `fit_one_model`, and `fit_models`
- [ ] All existing tests pass unchanged

## Future Work

- Benchmark on the spike analysis: compare Delta shift replicate correlation (R²) with and without `scale_fusion_by_n=True`
- Consider extending sample-size weighting to other regularization terms (l2reg, beta0_ridge) if the pattern proves valuable
- Investigate adaptive lasso (per-mutation weights) as a complementary approach
- Consider loss normalization (`.mean()` vs `.sum()`) as a separate enhancement
